### PR TITLE
Added Dialogs for direct Audiostream and Subtitle switching

### PR DESCRIPTION
**Pull request by lobermann**
## Description
Mentioned this in one of the last PRs I made. This Adds two new dialogs that allow direct switching of audiostream and subtitle.

## Motivation and Context
In order not to have to always navigate through the menu if you want to switch the audiostream or subtitle, this two dialogs can be called directly from the OSD and allow quick and easy change of them. (Similar to how it is on netflix for example).
Looking forward for feedback. Might not find it's way into Estuary (as we will not add any new buttons there) but there might be other skins interested in using it.

## How Has This Been Tested?
Tested on MacOS and LibreElec.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by lrusak** [comment]
Would you mind adding some screenshots?

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+#include "input/Key.h"
+#include "settings/Settings.h"
+#include "settings/lib/Setting.h"
+#include "utils/JobManager.h"
+#include "utils/LangCodeExpander.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
+#include "video/VideoDatabase.h"
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: I doubt that all these headers are actually needed. Example: `utils/log.h` included, but I cannot see any use of `CLog`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    return false;
+  }
+
+  CGUIDialogSelect* dialog =
+      CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogSelect>(
+          WINDOW_DIALOG_SELECT);
+  if (dialog == NULL)
+    return false;
+
+  CFileItemList options;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: I would define vars where they are actually first used.

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    CGUIDialogKaiToast::QueueNotification(
+        CGUIDialogKaiToast::Info, g_localizeStrings.Get(STRING_AUDIO_HEADER),
+        g_localizeStrings.Get(STRING_TOAST_TEXT_ERROR), TOAST_DISPLAY_TIME, false);
+    return false;
+  }
+
+  CGUIDialogSelect* dialog =
+      CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogSelect>(
+          WINDOW_DIALOG_SELECT);
+  if (dialog == NULL)
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: `if (dialog == nullptr)` or `if (!dialog)`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  CFileItemList options;
+
+  int audioStreamCount = g_application.GetAppPlayer().GetAudioStreamCount();
+  int currentAudio = g_application.GetAppPlayer().GetAudioStream();
+
+  // cycle through each subtitle and add it to our entry list
+  for (int i = 0; i < audioStreamCount; ++i)
+  {
+    std::string strItem;
+    std::string strLanguage;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: I would define vars where they are actually first used.

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    AudioStreamInfo info;
+    g_application.GetAppPlayer().GetAudioStreamInfo(i, info);
+
+    if (!g_LangCodeExpander.Lookup(info.language, strLanguage))
+      strLanguage = g_localizeStrings.Get(13205); // Unknown
+
+    if (info.name.length() == 0)
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: `if (info.name.empty())`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      strItem = strLanguage;
+    else
+      strItem = StringUtils::Format("%s - %s", strLanguage.c_str(), info.name.c_str());
+
+    CFileItemPtr item(new CFileItem(strItem.c_str()));
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: `std::make_shared<CFileItem>(strItem);`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  if (options.Size() < 2)
+  {
+    CGUIDialogKaiToast::QueueNotification(
+        CGUIDialogKaiToast::Info, g_localizeStrings.Get(STRING_AUDIO_HEADER),
+        g_localizeStrings.Get(STRING_TOAST_TEXT_ERROR), TOAST_DISPLAY_TIME, false);
+    return true;
+  }
+
+  dialog->Reset();
+  dialog->SetHeading(g_localizeStrings.Get(STRING_AUDIO_HEADER).c_str());
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: `.c_str()`not needed.

File: xbmc/video/dialogs/GUIDialogAudioSelect.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "guilib/GUIDialog.h"
+
+#include <string>
+
+class CFileItem;
+class CFileItemList;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.h]: These forward decls are not needed.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "guilib/GUIDialog.h"
+
+#include <string>
+
+class CFileItem;
+class CFileItemList;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.h]: These forward decls are not needed.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ *  Copyright (C) 2005-2019 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "guilib/GUIDialog.h"
+
+#include <string>
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.h]: include not needed.

File: xbmc/video/dialogs/GUIDialogAudioSelect.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ *  Copyright (C) 2005-2019 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "guilib/GUIDialog.h"
+
+#include <string>
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.h]: include not needed.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+#include "settings/lib/Setting.h"
+#include "utils/JobManager.h"
+#include "utils/LangCodeExpander.h"
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
+#include "video/VideoDatabase.h"
+
+#include "system.h"
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: I doubt that all these headers are actually needed. Example: `utils/log.h` included, but I cannot see any use of `CLog`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+#include "utils/StringUtils.h"
+#include "utils/URIUtils.h"
+#include "utils/Variant.h"
+#include "utils/log.h"
+#include "video/VideoDatabase.h"
+
+#include "system.h"
+
+using namespace ADDON;
+using namespace XFILE;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: I don't think you need this using statements.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+using namespace ADDON;
+using namespace XFILE;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: I don't think you need this using statements.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    return false;
+  }
+
+  CGUIDialogSelect* dialog =
+      CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogSelect>(
+          WINDOW_DIALOG_SELECT);
+  if (dialog == NULL)
+    return false;
+
+  CFileItemList options;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: I would define vars where they are actually first used.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  // cycle through each subtitle and add it to our entry list
+  for (int i = 0; i < subtitleStreamCount; ++i)
+  {
+    SubtitleStreamInfo info;
+    g_application.GetAppPlayer().GetSubtitleStreamInfo(i, info);
+
+    std::string strItem;
+    std::string strLanguage;
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: I would define vars where they are actually first used.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+  if (options.Size() < 2)
+  {
+    CGUIDialogKaiToast::QueueNotification(
+        CGUIDialogKaiToast::Info, g_localizeStrings.Get(STRING_SUBTITLES_HEADER),
+        g_localizeStrings.Get(STRING_TOAST_TEXT_ERROR), TOAST_DISPLAY_TIME, false);
+    return true;
+  }
+
+  dialog->Reset();
+  dialog->SetHeading(g_localizeStrings.Get(STRING_SUBTITLES_HEADER).c_str());
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: `.c_str()`not needed.

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+  CFileItemList options;
+
+  int subtitleStreamCount = g_application.GetAppPlayer().GetSubtitleCount();
+  int currentSubtitle = g_application.GetAppPlayer().GetSubtitle();
+  if (!g_application.GetAppPlayer().GetSubtitleVisible())
+    currentSubtitle = -1;
+
+  //Add the Disabled Element
+  {
+    CFileItemPtr item(new CFileItem(g_localizeStrings.Get(1223).c_str()));
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: `std::make_shared<CFileItem>(g_localizeStrings.Get(1223));`

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+    SubtitleStreamInfo info;
+    g_application.GetAppPlayer().GetSubtitleStreamInfo(i, info);
+
+    std::string strItem;
+    std::string strLanguage;
+
+    if (!g_LangCodeExpander.Lookup(info.language, strLanguage))
+      strLanguage = g_localizeStrings.Get(13205); // Unknown
+
+    if (info.name.length() == 0)
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: `if (info.name.empty())`

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+    CGUIDialogKaiToast::QueueNotification(
+        CGUIDialogKaiToast::Info, g_localizeStrings.Get(STRING_SUBTITLES_HEADER),
+        g_localizeStrings.Get(STRING_TOAST_TEXT_ERROR), TOAST_DISPLAY_TIME, false);
+    return false;
+  }
+
+  CGUIDialogSelect* dialog =
+      CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogSelect>(
+          WINDOW_DIALOG_SELECT);
+  if (dialog == NULL)
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: `if (dialog == nullptr)` or `if (!dialog)`

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+
+    if (!g_LangCodeExpander.Lookup(info.language, strLanguage))
+      strLanguage = g_localizeStrings.Get(13205); // Unknown
+
+    if (info.name.length() == 0)
+      strItem = strLanguage;
+    else
+      strItem = StringUtils::Format("%s - %s", strLanguage.c_str(), info.name.c_str());
+
+    CFileItemPtr item(new CFileItem(strItem.c_str()));
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: `std::make_shared<CFileItem>(strItem);`

File: xbmc/video/dialogs/GUIDialogAudioSelect.cpp [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+bool CGUIDialogAudioSelect::OnMessage(CGUIMessage& message)
+{
+  return CGUIDialog::OnMessage(message);
+}
+
+void CGUIDialogAudioSelect::OnInitWindow()
+{
+  ShowAudioSelect();
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogAudioSelect.cpp]: Are you sure it is correct to show the dialog in `OnInitWindow` - while the dialog is not yet fully initialized? Did you copy this pattern from an existing dialog implementation?

File: xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+bool CGUIDialogSubtitleSelect::OnMessage(CGUIMessage& message)
+{
+  return CGUIDialog::OnMessage(message);
+}
+
+void CGUIDialogSubtitleSelect::OnInitWindow()
+{
+  ShowSubtitleSelect();
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogSubtitleSelect.cpp]: Are you sure it is correct to show the dialog in OnInitWindow - while the dialog is not yet fully initialized? Did you copy this pattern from an existing dialog implementation?

**Review by ksooo — CHANGES_REQUESTED** [review]
This needs a bit more love and I'm not sure the implementation is correct at all. Implementing a dialog which does nothing than to open another dialog from its `OnInitWindow`override looks strange.

**Comment by da-anda** [comment]
not against such a dialog in general, but aren't skins already able to add such a dialog/feature using currently available infolabels etc? I remember the balloon OSD popup in Confluence that had a similar feature.

And since you mentined Netflix. Netflix uses a combined view to switch audio and subtitle language which I actually like quite a lot. What was your underlying idea for going with two separate dialogs instead of having something like a `LanguageSelectDialog` (or however we might call that)?

**Comment by lobermann** [comment]
@da-anda @ksooo to follow up on my open PR's, I guess we are not going to merge this one as it is and also it seems that this functionality is also achievable just within the skin. So I guess we can close this?

**Comment by da-anda** [comment]
strictly by looking at the code, I can't really say if it makes sense to add these dialogs or not. I know for certain that you can cycle through subs and audio streams via builtins that can be triggered by the skin in any way it likes. I however am not sure if skins are currently able to get a list of all available streams/languages for subs, audio tracks and video (in case of multi video stream containers) and then directly select a specific one. This in fact might be a nice addition to have (in case it's currently not yet possible), but I'm not sure if the best/only way to do this will be a new dialog that skins MUST add since it will be new core UI element/dialog that way, or if it would somehow be possible achieve this just by using infolabels and builtins, so that could integrate it in whatever form they like (new dialog or just via a menu that is tightly integrated into the OSD, like in Netflix).

@ronie do you know out of the top of your head if a feature like this would be possible via just infolabels and builtins?

**Comment by luzpaz** [comment]
any traction here ?

**Comment by Hitcher** [comment]
Having the ability to open these dialogs directly is a positive imo.

**Comment by jjd-uk** [comment]
Agreed it would be nice for the skin to be able to directly call the audio and subtitle selection dialogs. Would allow an improved version of https://github.com/xbmc/xbmc/pull/22557 where the toggle buttons are replaced by buttons opening the selection dialogs.

**Comment by CastagnaIT** [comment]
superseeded by #25762

**Pull request closed without merging**